pyvrp/cpp/educate/LocalSearch.cpp

```diff
@@ -172,6 +186,70 @@ bool LocalSearch::applyRouteOps(Route *U,
     return false;
 }
 
+void LocalSearch::maybeInsert(Node *U,
+                              Node *V,
+                              CostEvaluator const &costEvaluator)
+{
+    assert(!U->route && V->route);
+
+    auto const &uClient = data.client(U->client);
+
+    int const current = data.dist(V->client, n(V)->client);
+    int const proposed = data.dist(V->client, U->client)
+                         + data.dist(U->client, n(V)->client) - uClient.prize;
+
+    int deltaCost = proposed - current;
+
+    deltaCost += costEvaluator.loadPenalty(V->route->load() + uClient.demand,
+                                           data.vehicleCapacity());
+    deltaCost
+        -= costEvaluator.loadPenalty(V->route->load(), data.vehicleCapacity());
+
+    auto vTWS
+        = TWS::merge(data.durationMatrix(), V->twBefore, U->tw, n(V)->twAfter);
+
+    deltaCost += costEvaluator.twPenalty(vTWS.totalTimeWarp());
+    deltaCost -= costEvaluator.twPenalty(V->route->timeWarp());
+
+    if (deltaCost < 0)
+    {
+        U->insertAfter(V);           // U has no route, so there's nothing to
+        update(V->route, V->route);  // update there.
+    }
+}
+
+void LocalSearch::maybeRemove(Node *U, CostEvaluator const &costEvaluator)
+{
+    assert(U->route);
+
+    auto const &uClient = data.client(U->client);
+
+    int const current = data.dist(p(U)->client, U->client)
+                        + data.dist(U->client, n(U)->client) - uClient.prize;
+
+    int const proposed = data.dist(p(U)->client, n(U)->client);
+
+    int deltaCost = proposed - current;
+
+    deltaCost += costEvaluator.loadPenalty(U->route->load() - uClient.demand,
+                                           data.vehicleCapacity());
+    deltaCost
+        -= costEvaluator.loadPenalty(U->route->load(), data.vehicleCapacity());
+
```

Comment:
We could add a precheck here
```suggestion
    if (deltaCost - U->route->timeWarp() >= 0)
        return deltaCost
```